test/cpp/ext/otel/otel_plugin_test.cc

```diff
@@ -38,6 +43,73 @@ namespace grpc {
 namespace testing {
 namespace {
 
+#define GRPC_ARG_SERVER_SELECTOR_KEY "grpc.testing.server_selector_key"
+#define GRPC_ARG_SERVER_SELECTOR_VALUE "grpc.testing.server_selector_value"
+
+template <typename T>
+void PopulateLabelMap(
+    T label_keys, T label_values,
+    std::unordered_map<std::string,
+                       opentelemetry::sdk::common::OwnedAttributeValue>*
+        label_maps) {
+  for (int i = 0; i < label_keys.size(); ++i) {
+    (*label_maps)[std::string(label_keys[i])] = std::string(label_values[i]);
+  }
+}
+
+MATCHER_P4(AttributesEq, label_keys, label_values, optional_label_keys,
+           optional_label_values, "") {
+  std::unordered_map<std::string,
+                     opentelemetry::sdk::common::OwnedAttributeValue>
+      label_map;
+  PopulateLabelMap(label_keys, label_values, &label_map);
+  PopulateLabelMap(optional_label_keys, optional_label_values, &label_map);
+  return ::testing::ExplainMatchResult(
+      ::testing::UnorderedElementsAreArray(label_map),
+      arg.attributes.GetAttributes(), result_listener);
+}
+
+template <typename T>
+auto IntOrDoubleEq(T result) {
+  return ::testing::Eq(result);
+}
+template <>
+auto IntOrDoubleEq(double result) {
+  return ::testing::DoubleEq(result);
+}
+
+MATCHER_P(CounterResultEq, result, "") {
+  return ::testing::ExplainMatchResult(
+      ::testing::VariantWith<opentelemetry::sdk::metrics::SumPointData>(
+          ::testing::Field(
+              &opentelemetry::sdk::metrics::SumPointData::value_,
+              ::testing::VariantWith<std::remove_cv_t<decltype(result)>>(
+                  IntOrDoubleEq(result)))),
+      arg.point_data, result_listener);
+}
+
+MATCHER_P4(HistogramResultEq, sum, min, max, count, "") {
```

Comment:
I'm not a fan of complicated matchers, but since this is already written, I'm fine with it remaining. My only concern is making debugging easier. My experience with such matchers is that on errors, it is really hard to figure out what went wrong, so at the very minimum, we need to print out what the actual result was so we can compare it to the expectations.

Reply:
I see, let me test it out